docs/upgrade-to-v6.md

```diff
@@ -0,0 +1,507 @@
+```java
+RedirectHandlerOption redirectHandlerOption = new RedirectHandlerOption(5, response -> {
+    return response.code() == 200 ? true : false;} // Only redirect if the response code is 200
+);
+graphClient.me().get(requestConfiguration -> {
+    requestConfiguration.options.add(redirectHandlerOption);
+});
+```
+
+Current middleware enabled by default includes: 
+- `RedirectHandler` 
+- `RetryHandler` 
+- `ParametersNameDecodingHandler` 
+- `UserAgentHandler`
+- `HeadersInspectionHandler`
+
+`RequestOption` implementations for these handlers can be found [here](https://github.com/microsoft/kiota-java/tree/main/components/http/okHttp/src/main/java/com/microsoft/kiota/http/middleware/options).
+
+>NOTE: Users will need to add the following dependency to their project in order to make use of the currently available option implementations:
+`implementation 'com.microsoft.kiota:microsoft-kiota-http-okHttp:0.12.1'` 
+For latest version see the [Kiota-Java Repo](https://github.com/microsoft/kiota-java)
+
+### Native Response Object
+
+Users can pass an instance of the `NativeResponseHandler` class, an implementation of the `ResponseHandler` interface, within a per-request `ResponseHandlerOption` to get a native `Okhttp3.Response` object from a request. This can be used in scenarios where a user wants to access the native response object and it's properties directly. 
+
+In the following example, the `NativeResponseHandler` is passed as a property of the `ResponseHandlerOption` class to the `RequestConfiguration` lambda. The `NativeResponseHandler` is then used to get the native response object after the request has been made. 
+```java
+NativeResponseHandler nativeResponseHandler = new NativeResponseHandler();
+ResponseHandlerOption responseHandlerOption = new ResponseHandlerOption();
+responseHandlerOption.setResponseHandler(nativeResponseHandler);
+graphClient.me().get(requestConfiguration -> {
+    requestConfiguration.options.add(responseHandlerOption);
+});
+Response nativeResponse = (Response) nativeResponseHandler.getValue();
+```
+Users can also fully customize response handling 
+
+### Support for OData Casts
+
+OData allows downcasting the result to the right specialized type, meaning that users can have access to all properties in the response, rather than pulling those out from the AdditionalData holder. The fluent API pattern now comes with enriched segments allowing users to request a specific Odata type in the event that an endpoint supports multiple types. 
+
+For example:
+```java
+// Get the collection of Users in a group
+// API Endpoint: /groups/{id}/members/microsoft.graph.user
+UserCollectionResponse users = graphClient.groups().byGroupId("Group-Id").members().graphUser().get();
+// Get the collection of Applications in a group
+// API Endpoint: /groups/{id}/members/microsoft.graph.application
+ApplicationCollectionResponse applications = graphClient.groups().byGroupId("Group-Id").members().graphApplication().get();
```

Comment:
the raw count is missing

Reply:
Do you mean: 
graphClient.groups().byGroupId("").members().get()?

Reply:
no I meant the new endpoints we support like `graphClient.groups().count()`. This SDK didn't use to support it.

Reply:
The SDK as of v5 has supported count requests. 
```graphClient.groups().count().buildRequest();``` this is a valid request in v5

Reply:
I think the surface area of where .count() is supported has expanded, but I dont believe .count() is a new new feature. I can add a note to explain that the support has increased which I agree is important. Where do you think would be a good place to add the note or even section?

Reply:
Really? I thought we didn't have them for dotnet/java with typewriter. Ok disregard this comment then